tests/chainer_tests/links_tests/normalization_tests/test_batch_normalization.py

```diff
@@ -37,118 +29,206 @@ def _batch_normalization(expander, gamma, beta, x, mean, var, eps, test):
+            var = x.var(axis=self.aggr_axes, keepdims=True)
+            std = numpy.sqrt(var + self.eps)
+        y = gamma[self.expander] * (x - mean) / std + beta[self.expander]
+        return y,
 
-    @condition.retry(3)
-    def test_forward_cpu(self):
-        self.check_forward(self.x)
 
-    @attr.gpu
-    @condition.retry(3)
-    def test_forward_gpu(self):
-        self.link.to_gpu()
-        self.check_forward(cuda.to_gpu(self.x))
+@_inject_backend_tests
+@_parameterize
+class BatchNormalizationTest(BatchNormalizationTestBase, testing.LinkTestCase):
 
-    @attr.cudnn
-    def test_forward_gpu_without_cudnn(self):
-        with chainer.using_config('use_cudnn', 'never'):
-            self.test_forward_gpu()
+    pass
 
-    @attr.multi_gpu(2)
-    @condition.retry(3)
-    def test_forward_multi_gpu(self):
-        with cuda.get_device_from_id(1):
-            self.link.to_gpu()
-            x = cuda.to_gpu(self.x)
-        with cuda.get_device_from_id(0):
-            self.check_forward(x)
 
-    def check_backward(self, x_data, y_grad):
-        gradient_check.check_backward(
-            self.link, x_data, y_grad, (self.link.gamma, self.link.beta),
-            eps=1e-2, **self.check_backward_optionss)
+# TODO(hvy): Safely remove this test class when BackendConfig no longer
+# modifies the current device since those cases should be covered by the tests
+# above.
+@attr.multi_gpu(2)
+@testing.inject_backend_tests(
+    None,
+    testing.product({
+        'use_cuda': [True],
+        'use_cudnn': ['never', 'always'],
+        'cuda_device': [0],
+    }))
+@_parameterize
+class BatchNormalizationMultiGpuTest(
+        BatchNormalizationTestBase, testing.LinkTestCase):
 
-    @condition.retry(3)
-    def test_backward_cpu(self):
-        self.check_backward(self.x, self.gy)
+    skip_backward_test = True
 
-    @attr.gpu
-    @condition.retry(3)
-    def test_backward_gpu(self):
-        self.link.to_gpu()
-        self.check_backward(cuda.to_gpu(self.x), cuda.to_gpu(self.gy))
+    # TODO(hvy): Remove this relaxation. It is currently needed as the
+    # inter-device copy in CuPy with non-contiguous arrays are broken.
+    contiguous = 'C'
 
-    @attr.cudnn
-    def test_backward_gpu_without_cudnn(self):
-        with chainer.using_config('use_cudnn', 'never'):
-            self.test_backward_gpu()
+    def forward(self, link, inputs, device):
+        x, = inputs
+
+        device_1 = backend.GpuDevice.from_device_id(1)
+        link.to_device(device_1)
+        x.to_device(device_1)
+
+        device_0 = backend.GpuDevice.from_device_id(0)
+        with chainer.using_device(device_0):
+            with chainer.using_config('train', not self.test):
+                y = link(x, finetune=self.finetune)
+        return y,
+
+
+@testing.parameterize(*(testing.product_dict(
+    testing.product({
+        'dtype': [numpy.float16, numpy.float32, numpy.float64],
+    }))))
+@_inject_backend_tests
```

Comment:
Needs discussion: whether to parameterize backends in initializers test.

Reply:
Leaving it as is according to offline discussions. https://github.com/chainer/chainer/pull/6187#issuecomment-471829926